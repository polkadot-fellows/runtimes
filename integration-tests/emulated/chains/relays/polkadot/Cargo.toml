[package]
name = "polkadot-emulated-chain"
version.workspace = true
authors.workspace = true
edition.workspace = true
license = "Apache-2.0"
description = "Polkadot emulated chain used for integration tests"
publish = false

[dependencies]

# Substrate
sp-core = { workspace = true, default-features = true }
sp-runtime = { workspace = true, default-features = true }
authority-discovery-primitives = { package = "sp-authority-discovery", version = "27.0.0" }
babe-primitives = { package = "sp-consensus-babe", version = "0.33.0" }
beefy-primitives = { package = "sp-consensus-beefy", version = "14.0.0" }
grandpa = { package = "sc-consensus-grandpa", version = "0.20.0" }
pallet-staking = { workspace = true, default-features = true }

# Polkadot
polkadot-primitives = { workspace = true, default-features = true }

# Cumulus
parachains-common = { workspace = true, default-features = true }
emulated-integration-tests-common = { workspace = true }

# Runtimes
polkadot-runtime-constants = { path = "../../../../../relay/polkadot/constants" }
polkadot-runtime = { path = "../../../../../relay/polkadot" }
