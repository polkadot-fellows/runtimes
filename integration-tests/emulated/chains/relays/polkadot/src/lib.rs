// Copyright (C) Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub mod genesis;

// Cumulus
use emulated_integration_tests_common::{
	impl_accounts_helpers_for_relay_chain, impl_assert_events_helpers_for_relay_chain,
	impl_hrmp_channels_helpers_for_relay_chain, impl_send_transact_helpers_for_relay_chain,
	xcm_emulator::decl_test_relay_chains,
};

// Polkadot declaration
decl_test_relay_chains! {
	#[api_version(11)]
	pub struct Polkadot {
		genesis = genesis::genesis(),
		on_init = (),
		runtime = polkadot_runtime,
		core = {
			SovereignAccountOf: polkadot_runtime::xcm_config::SovereignAccountOf,
		},
		pallets = {
			XcmPallet: polkadot_runtime::XcmPallet,
			Balances: polkadot_runtime::Balances,
			Treasury: polkadot_runtime::Treasury,
			AssetRate: polkadot_runtime::AssetRate,
			Hrmp: polkadot_runtime::Hrmp,
			Identity: polkadot_runtime::Identity,
			IdentityMigrator: polkadot_runtime::IdentityMigrator,
		}
	},
}

// Polkadot implementation
impl_accounts_helpers_for_relay_chain!(Polkadot);
impl_assert_events_helpers_for_relay_chain!(Polkadot);
impl_hrmp_channels_helpers_for_relay_chain!(Polkadot);
impl_send_transact_helpers_for_relay_chain!(Polkadot);
