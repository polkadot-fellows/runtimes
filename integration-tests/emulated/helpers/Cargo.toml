[package]
name = "integration-tests-helpers"
version.workspace = true
authors.workspace = true
edition.workspace = true
license = "Apache-2.0"
description = "Emulated integration tests helpers"
publish = false

[dependencies]
paste = { workspace = true }

# Substrate
pallet-balances = { workspace = true, default-features = true }
pallet-message-queue = { workspace = true, default-features = true }

# Polkadot
xcm = { package = "staging-xcm", version = "8.0.1" }
pallet-xcm = { workspace = true, default-features = true }

# Cumulus
xcm-emulator = { workspace = true }
cumulus-pallet-xcmp-queue = { workspace = true, default-features = true }
asset-test-utils = { workspace = true }
