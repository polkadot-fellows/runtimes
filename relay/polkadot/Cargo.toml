[package]
name = "polkadot-runtime"
build = "build.rs"
repository.workspace = true
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
bitvec = { version = "1.0.0", default-features = false, features = ["alloc"] }
parity-scale-codec = { version = "3.6.1", default-features = false, features = ["derive", "max-encoded-len"] }
scale-info = { version = "2.5.0", default-features = false, features = ["derive"] }
log = { version = "0.4.17", default-features = false }
rustc-hex = { version = "2.1.0", default-features = false }
serde = { version = "1.0.188", default-features = false }
serde_derive = { version = "1.0.117", optional = true }
static_assertions = "1.1.0"
smallvec = "1.8.0"

authority-discovery-primitives = { package = "sp-authority-discovery", default-features = false , version = "21.0.0" }
babe-primitives = { package = "sp-consensus-babe", default-features = false , version = "0.27.0" }
beefy-primitives = { package = "sp-consensus-beefy", default-features = false , version = "8.0.0" }
block-builder-api = { package = "sp-block-builder", default-features = false , version = "21.0.0" }
inherents = { package = "sp-inherents", default-features = false , version = "21.0.0" }
offchain-primitives = { package = "sp-offchain", default-features = false , version = "21.0.0" }
tx-pool-api = { package = "sp-transaction-pool", default-features = false , version = "21.0.0" }
sp-arithmetic = { default-features = false , version = "18.0.0" }
sp-api = { default-features = false , version = "21.0.0" }
sp-std = { default-features = false , version = "10.0.0" }
sp-io = { default-features = false , version = "25.0.0" }
sp-mmr-primitives = { default-features = false , version = "21.0.0" }
sp-runtime = { default-features = false , version = "26.0.0" }
sp-staking = { default-features = false , version = "21.0.0" }
sp-core = { default-features = false , version = "23.0.0" }
sp-session = { default-features = false , version = "22.0.0" }
sp-storage = { default-features = false , version = "15.0.0" }
sp-version = { default-features = false , version = "24.0.0" }
sp-npos-elections = { default-features = false , version = "21.0.0" }

pallet-authority-discovery = { default-features = false , version = "23.0.0" }
pallet-authorship = { default-features = false , version = "23.0.0" }
pallet-babe = { default-features = false , version = "23.0.0" }
pallet-bags-list = { default-features = false , version = "22.0.0" }
pallet-balances = { default-features = false , version = "23.0.0" }
pallet-bounties = { default-features = false , version = "22.0.0" }
pallet-child-bounties = { default-features = false , version = "22.0.0" }
pallet-transaction-payment = { default-features = false , version = "23.0.0" }
pallet-transaction-payment-rpc-runtime-api = { default-features = false , version = "23.0.0" }
pallet-collective = { default-features = false , version = "23.0.0" }
pallet-conviction-voting = { default-features = false , version = "23.0.0" }
pallet-democracy = { default-features = false , version = "23.0.0" }
pallet-elections-phragmen = { default-features = false , version = "24.0.0" }
pallet-election-provider-multi-phase = { default-features = false , version = "22.0.0" }
pallet-fast-unstake = { default-features = false , version = "22.0.0" }
frame-executive = { default-features = false , version = "23.0.0" }
pallet-grandpa = { default-features = false , version = "23.0.0" }
pallet-identity = { default-features = false , version = "23.0.0" }
pallet-im-online = { default-features = false , version = "22.0.0" }
pallet-indices = { default-features = false , version = "23.0.0" }
pallet-membership = { default-features = false , version = "23.0.0" }
pallet-message-queue = { default-features = false , version = "26.0.0" }
pallet-multisig = { default-features = false , version = "23.0.0" }
pallet-nomination-pools = { default-features = false , version = "20.0.0" }
pallet-nomination-pools-runtime-api = { default-features = false , version = "18.0.0" }
pallet-offences = { default-features = false , version = "22.0.0" }
pallet-preimage = { default-features = false , version = "23.0.0" }
pallet-proxy = { default-features = false , version = "23.0.0" }
pallet-referenda = { default-features = false , version = "23.0.0" }
pallet-scheduler = { default-features = false , version = "24.0.0" }
pallet-session = { default-features = false , version = "23.0.0" }
frame-support = { default-features = false , version = "23.0.0" }
pallet-staking = { default-features = false , version = "23.0.0" }
pallet-staking-reward-curve = { version = "8.0.0" }
pallet-staking-runtime-api = { default-features = false , version = "9.0.0" }
frame-system = { default-features = false , version = "23.0.0" }
frame-system-rpc-runtime-api = { default-features = false , version = "21.0.0" }
polkadot-runtime-constants = { package = "polkadot-runtime-constants", path = "constants", default-features = false }
pallet-timestamp = { default-features = false , version = "22.0.0" }
pallet-tips = { default-features = false , version = "22.0.0" }
pallet-treasury = { default-features = false , version = "22.0.0" }
pallet-whitelist = { default-features = false , version = "22.0.0" }
pallet-vesting = { default-features = false , version = "23.0.0" }
pallet-utility = { default-features = false , version = "23.0.0" }
frame-election-provider-support = { default-features = false , version = "23.0.0" }
pallet-xcm = { default-features = false, features=["experimental"] , version = "2.0.0" }
pallet-xcm-benchmarks = { default-features = false, optional = true , version = "2.0.0" }

frame-benchmarking = { default-features = false, optional = true , version = "23.0.0" }
frame-try-runtime = { default-features = false, optional = true , version = "0.29.0" }
frame-system-benchmarking = { default-features = false, optional = true , version = "23.0.0" }
pallet-election-provider-support-benchmarking = { default-features = false, optional = true , version = "22.0.0" }
pallet-offences-benchmarking = { default-features = false, optional = true , version = "23.0.0" }
pallet-session-benchmarking = { default-features = false, optional = true , version = "23.0.0" }
pallet-nomination-pools-benchmarking = { default-features = false, optional = true , version = "21.0.0" }
hex-literal = { version = "0.4.1", optional = true }

runtime-common = { package = "polkadot-runtime-common", default-features = false, features = ["experimental"] , version = "2.0.0" }
runtime-parachains = { package = "polkadot-runtime-parachains", default-features = false , version = "2.0.0" }
primitives = { package = "polkadot-primitives", default-features = false , version = "2.0.0" }

xcm = { package = "staging-xcm", default-features = false , version = "2.0.0" }
xcm-executor = { package = "staging-xcm-executor", default-features = false , version = "2.0.0" }
xcm-builder = { package = "staging-xcm-builder", default-features = false , version = "2.0.0" }

[dev-dependencies]
hex-literal = "0.4.1"
tiny-keccak = { version = "2.0.2", features = ["keccak"] }
keyring = { package = "sp-keyring", version = "26.0.0" }
sp-trie = { version = "24.0.0" }
serde_json = "1.0.96"
separator = "0.4.1"
remote-externalities = { package = "frame-remote-externalities" , version = "0.30.0" }
tokio = { version = "1.24.2", features = ["macros"] }
sp-tracing = { default-features = false , version = "12.0.0" }

[build-dependencies]
substrate-wasm-builder = { version = "12.0.0" }

[features]
default = [ "std" ]
no_std = []
only-staking = []
std = [
	"authority-discovery-primitives/std",
	"babe-primitives/std",
	"beefy-primitives/std",
	"bitvec/std",
	"block-builder-api/std",
	"frame-benchmarking?/std",
	"frame-election-provider-support/std",
	"frame-executive/std",
	"frame-support/std",
	"frame-system-benchmarking?/std",
	"frame-system-rpc-runtime-api/std",
	"frame-system/std",
	"frame-try-runtime/std",
	"inherents/std",
	"log/std",
	"offchain-primitives/std",
	"pallet-authority-discovery/std",
	"pallet-authorship/std",
	"pallet-babe/std",
	"pallet-bags-list/std",
	"pallet-balances/std",
	"pallet-bounties/std",
	"pallet-child-bounties/std",
	"pallet-collective/std",
	"pallet-conviction-voting/std",
	"pallet-democracy/std",
	"pallet-election-provider-multi-phase/std",
	"pallet-election-provider-support-benchmarking?/std",
	"pallet-elections-phragmen/std",
	"pallet-fast-unstake/std",
	"pallet-grandpa/std",
	"pallet-identity/std",
	"pallet-im-online/std",
	"pallet-indices/std",
	"pallet-membership/std",
	"pallet-message-queue/std",
	"pallet-multisig/std",
	"pallet-nomination-pools-benchmarking?/std",
	"pallet-nomination-pools-runtime-api/std",
	"pallet-nomination-pools/std",
	"pallet-offences-benchmarking?/std",
	"pallet-offences/std",
	"pallet-preimage/std",
	"pallet-proxy/std",
	"pallet-referenda/std",
	"pallet-scheduler/std",
	"pallet-session-benchmarking?/std",
	"pallet-session/std",
	"pallet-staking-runtime-api/std",
	"pallet-staking/std",
	"pallet-timestamp/std",
	"pallet-tips/std",
	"pallet-transaction-payment-rpc-runtime-api/std",
	"pallet-transaction-payment/std",
	"pallet-treasury/std",
	"pallet-utility/std",
	"pallet-vesting/std",
	"pallet-whitelist/std",
	"pallet-xcm-benchmarks?/std",
	"pallet-xcm/std",
	"parity-scale-codec/std",
	"polkadot-runtime-constants/std",
	"primitives/std",
	"runtime-common/std",
	"runtime-parachains/std",
	"rustc-hex/std",
	"scale-info/std",
	"serde/std",
	"serde_derive",
	"sp-api/std",
	"sp-arithmetic/std",
	"sp-core/std",
	"sp-io/std",
	"sp-mmr-primitives/std",
	"sp-npos-elections/std",
	"sp-runtime/std",
	"sp-session/std",
	"sp-staking/std",
	"sp-std/std",
	"sp-storage/std",
	"sp-tracing/std",
	"sp-version/std",
	"tx-pool-api/std",
	"xcm-builder/std",
	"xcm-executor/std",
	"xcm/std",
]
runtime-benchmarks = [
	"frame-benchmarking/runtime-benchmarks",
	"frame-election-provider-support/runtime-benchmarks",
	"frame-support/runtime-benchmarks",
	"frame-system-benchmarking/runtime-benchmarks",
	"frame-system/runtime-benchmarks",
	"hex-literal",
	"pallet-babe/runtime-benchmarks",
	"pallet-bags-list/runtime-benchmarks",
	"pallet-balances/runtime-benchmarks",
	"pallet-bounties/runtime-benchmarks",
	"pallet-child-bounties/runtime-benchmarks",
	"pallet-collective/runtime-benchmarks",
	"pallet-conviction-voting/runtime-benchmarks",
	"pallet-democracy/runtime-benchmarks",
	"pallet-election-provider-multi-phase/runtime-benchmarks",
	"pallet-election-provider-support-benchmarking/runtime-benchmarks",
	"pallet-elections-phragmen/runtime-benchmarks",
	"pallet-fast-unstake/runtime-benchmarks",
	"pallet-grandpa/runtime-benchmarks",
	"pallet-identity/runtime-benchmarks",
	"pallet-im-online/runtime-benchmarks",
	"pallet-indices/runtime-benchmarks",
	"pallet-membership/runtime-benchmarks",
	"pallet-message-queue/runtime-benchmarks",
	"pallet-multisig/runtime-benchmarks",
	"pallet-nomination-pools-benchmarking/runtime-benchmarks",
	"pallet-nomination-pools/runtime-benchmarks",
	"pallet-offences-benchmarking/runtime-benchmarks",
	"pallet-offences/runtime-benchmarks",
	"pallet-preimage/runtime-benchmarks",
	"pallet-proxy/runtime-benchmarks",
	"pallet-referenda/runtime-benchmarks",
	"pallet-scheduler/runtime-benchmarks",
	"pallet-session-benchmarking/runtime-benchmarks",
	"pallet-staking/runtime-benchmarks",
	"pallet-timestamp/runtime-benchmarks",
	"pallet-tips/runtime-benchmarks",
	"pallet-treasury/runtime-benchmarks",
	"pallet-utility/runtime-benchmarks",
	"pallet-vesting/runtime-benchmarks",
	"pallet-whitelist/runtime-benchmarks",
	"pallet-xcm-benchmarks/runtime-benchmarks",
	"pallet-xcm/runtime-benchmarks",
	"primitives/runtime-benchmarks",
	"runtime-common/runtime-benchmarks",
	"runtime-parachains/runtime-benchmarks",
	"sp-runtime/runtime-benchmarks",
	"sp-staking/runtime-benchmarks",
	"xcm-builder/runtime-benchmarks",
	"xcm-executor/runtime-benchmarks",
]
try-runtime = [
	"frame-election-provider-support/try-runtime",
	"frame-executive/try-runtime",
	"frame-support/try-runtime",
	"frame-system/try-runtime",
	"frame-try-runtime",
	"frame-try-runtime/try-runtime",
	"pallet-authority-discovery/try-runtime",
	"pallet-authorship/try-runtime",
	"pallet-babe/try-runtime",
	"pallet-bags-list/try-runtime",
	"pallet-balances/try-runtime",
	"pallet-bounties/try-runtime",
	"pallet-child-bounties/try-runtime",
	"pallet-collective/try-runtime",
	"pallet-conviction-voting/try-runtime",
	"pallet-democracy/try-runtime",
	"pallet-election-provider-multi-phase/try-runtime",
	"pallet-elections-phragmen/try-runtime",
	"pallet-fast-unstake/try-runtime",
	"pallet-grandpa/try-runtime",
	"pallet-identity/try-runtime",
	"pallet-im-online/try-runtime",
	"pallet-indices/try-runtime",
	"pallet-membership/try-runtime",
	"pallet-message-queue/try-runtime",
	"pallet-multisig/try-runtime",
	"pallet-nomination-pools/try-runtime",
	"pallet-offences/try-runtime",
	"pallet-preimage/try-runtime",
	"pallet-proxy/try-runtime",
	"pallet-referenda/try-runtime",
	"pallet-scheduler/try-runtime",
	"pallet-session/try-runtime",
	"pallet-staking/try-runtime",
	"pallet-timestamp/try-runtime",
	"pallet-tips/try-runtime",
	"pallet-transaction-payment/try-runtime",
	"pallet-treasury/try-runtime",
	"pallet-utility/try-runtime",
	"pallet-vesting/try-runtime",
	"pallet-whitelist/try-runtime",
	"pallet-xcm/try-runtime",
	"runtime-common/try-runtime",
	"runtime-parachains/try-runtime",
	"sp-runtime/try-runtime",
]

# A feature that should be enabled when the runtime should be build for on-chain
# deployment. This will disable stuff that shouldn't be part of the on-chain wasm
# to make it smaller like logging for example.
on-chain-release-build = [ "sp-api/disable-logging" ]

# Set timing constants (e.g. session period) to faster versions to speed up testing.
fast-runtime = []

runtime-metrics = [ "runtime-parachains/runtime-metrics", "sp-io/with-tracing" ]
