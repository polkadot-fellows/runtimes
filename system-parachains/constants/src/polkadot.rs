// Copyright (C) Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// Universally recognized accounts.
pub mod account {
	use frame_support::PalletId;

	/// Polkadot treasury pallet id, used to convert into AccountId
	pub const POLKADOT_TREASURY_PALLET_ID: PalletId = PalletId(*b"py/trsry");
	/// Alliance pallet ID.
	/// Used as a temporary place to deposit a slashed imbalance before teleporting to the Treasury.
	pub const ALLIANCE_PALLET_ID: PalletId = PalletId(*b"py/allia");
	/// Referenda pallet ID.
	/// Used as a temporary place to deposit a slashed imbalance before teleporting to the Treasury.
	pub const REFERENDA_PALLET_ID: PalletId = PalletId(*b"py/refer");
	/// Ambassador Referenda pallet ID.
	/// Used as a temporary place to deposit a slashed imbalance before teleporting to the Treasury.
	pub const AMBASSADOR_REFERENDA_PALLET_ID: PalletId = PalletId(*b"py/amref");
	/// Identity pallet ID.
	/// Used as a temporary place to deposit a slashed imbalance before teleporting to the Treasury.
	pub const IDENTITY_PALLET_ID: PalletId = PalletId(*b"py/ident");
	/// Fellowship treasury pallet ID
	pub const FELLOWSHIP_TREASURY_PALLET_ID: PalletId = PalletId(*b"py/feltr");
	/// Ambassador treasury pallet ID
	pub const AMBASSADOR_TREASURY_PALLET_ID: PalletId = PalletId(*b"py/ambtr");
}

/// Consensus-related.
pub mod consensus {
	/// Maximum number of blocks simultaneously accepted by the Runtime, not yet included
	/// into the relay chain.
	pub const UNINCLUDED_SEGMENT_CAPACITY: u32 = 1;
	/// How many parachain blocks are processed by the relay chain per parent. Limits the
	/// number of blocks authored per slot.
	pub const BLOCK_PROCESSING_VELOCITY: u32 = 1;
	/// Relay chain slot duration, in milliseconds.
	pub const RELAY_CHAIN_SLOT_DURATION_MILLIS: u32 = 6000;
}

/// Constants relating to DOT.
pub mod currency {
	use polkadot_core_primitives::Balance;

	/// The default existential deposit for system chains. 1/10th of the Relay Chain's existential
	/// deposit. Individual system parachains may modify this in special cases.
	pub const SYSTEM_PARA_EXISTENTIAL_DEPOSIT: Balance =
		polkadot_runtime_constants::currency::EXISTENTIAL_DEPOSIT / 10;

	/// One "DOT" that a UI would show a user.
	pub const UNITS: Balance = 10_000_000_000;
	pub const DOLLARS: Balance = UNITS; // 10_000_000_000
	pub const GRAND: Balance = DOLLARS * 1_000; // 10_000_000_000_000
	pub const CENTS: Balance = DOLLARS / 100; // 100_000_000
	pub const MILLICENTS: Balance = CENTS / 1_000; // 100_000

	/// Deposit rate for stored data. 1/100th of the Relay Chain's deposit rate. `items` is the
	/// number of keys in storage and `bytes` is the size of the value.
	pub const fn system_para_deposit(items: u32, bytes: u32) -> Balance {
		polkadot_runtime_constants::currency::deposit(items, bytes) / 100
	}
}

/// Constants related to Polkadot fee payment.
pub mod fee {
	use frame_support::{
		pallet_prelude::Weight,
		weights::{
			constants::ExtrinsicBaseWeight, FeePolynomial, WeightToFeeCoefficient,
			WeightToFeeCoefficients, WeightToFeePolynomial,
		},
	};
	use polkadot_core_primitives::Balance;
	use smallvec::smallvec;
	pub use sp_runtime::Perbill;

	/// The block saturation level. Fees will be updates based on this value.
	pub const TARGET_BLOCK_FULLNESS: Perbill = Perbill::from_percent(25);

	/// Cost of every transaction byte at Polkadot system parachains.
	///
	/// It is the Relay Chain (Polkadot) `TransactionByteFee` / 10.
	pub const TRANSACTION_BYTE_FEE: Balance = super::currency::MILLICENTS;

	/// Handles converting a weight scalar to a fee value, based on the scale and granularity of the
	/// node's balance type.
	///
	/// This should typically create a mapping between the following ranges:
	///   - [0, MAXIMUM_BLOCK_WEIGHT]
	///   - [Balance::min, Balance::max]
	///
	/// Yet, it can be used for any other sort of change to weight-fee. Some examples being:
	///   - Setting it to `0` will essentially disable the weight fee.
	///   - Setting it to `1` will cause the literal `#[weight = x]` values to be charged.
	pub struct WeightToFee;
	impl frame_support::weights::WeightToFee for WeightToFee {
		type Balance = Balance;

		fn weight_to_fee(weight: &Weight) -> Self::Balance {
			let time_poly: FeePolynomial<Balance> = RefTimeToFee::polynomial().into();
			let proof_poly: FeePolynomial<Balance> = ProofSizeToFee::polynomial().into();

			// Take the maximum instead of the sum to charge by the more scarce resource.
			time_poly.eval(weight.ref_time()).max(proof_poly.eval(weight.proof_size()))
		}
	}

	/// Maps the reference time component of `Weight` to a fee.
	pub struct RefTimeToFee;
	impl WeightToFeePolynomial for RefTimeToFee {
		type Balance = Balance;
		fn polynomial() -> WeightToFeeCoefficients<Self::Balance> {
			// In Polkadot, extrinsic base weight (smallest non-zero weight) is mapped to 1/10 CENT:
			// The standard system parachain configuration is 1/10 of that, as in 1/100 CENT.
			let p = super::currency::CENTS;
			let q = 100 * Balance::from(ExtrinsicBaseWeight::get().ref_time());

			smallvec![WeightToFeeCoefficient {
				degree: 1,
				negative: false,
				coeff_frac: Perbill::from_rational(p % q, q),
				coeff_integer: p / q,
			}]
		}
	}

	/// Maps the proof size component of `Weight` to a fee.
	pub struct ProofSizeToFee;
	impl WeightToFeePolynomial for ProofSizeToFee {
		type Balance = Balance;
		fn polynomial() -> WeightToFeeCoefficients<Self::Balance> {
			// Map 10kb proof to 1 CENT.
			let p = super::currency::CENTS;
			let q = 10_000;

			smallvec![WeightToFeeCoefficient {
				degree: 1,
				negative: false,
				coeff_frac: Perbill::from_rational(p % q, q),
				coeff_integer: p / q,
			}]
		}
	}
}
