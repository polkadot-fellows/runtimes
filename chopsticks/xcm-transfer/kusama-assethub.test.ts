import { query, tx } from '../../helpers/api'

import { kusama } from '../../networks/polkadot'
import { statemine } from '../../networks/statemint'

import buildTest from './shared'

const tests = [
  {
    from: 'kusama',
    to: 'statemine',
    name: 'KSM',
    test: {
      xcmPalletDown: {
        tx: tx.xcmPallet.limitedReserveTransferAssetsV3(kusama.ksm, 1e12, tx.xcmPallet.parachainV3(0, statemine.paraId)),
        balance: query.balances,
      },
    },
  },
] as const


export type TestType = (typeof tests)[number]

buildTest(tests)
