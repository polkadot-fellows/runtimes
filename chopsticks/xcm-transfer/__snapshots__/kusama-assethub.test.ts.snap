// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'kusama' -> 'statemine' xcm transfer 'KSM' > xcmPallet transfer > balance on from chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 9000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 1,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`'kusama' -> 'statemine' xcm transfer 'KSM' > xcmPallet transfer > balance on to chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": 0,
    "free": 10000000000000,
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`'kusama' -> 'statemine' xcm transfer 'KSM' > xcmPallet transfer > to chain dmp events 1`] = `
[
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "outcome": {
        "Incomplete": [
          {
            "proofSize": 0,
            "refTime": 1000000000,
          },
          "UntrustedReserveLocation",
        ],
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
  {
    "data": {
      "dmqHead": "(hash)",
      "weightUsed": {
        "proofSize": 0,
        "refTime": 1000000000,
      },
    },
    "method": "DownwardMessagesProcessed",
    "section": "parachainSystem",
  },
]
`;

exports[`'kusama' -> 'statemine' xcm transfer 'KSM' > xcmPallet transfer > tx events 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": "(rounded 753000000)",
        },
      },
    },
    "method": "Attempted",
    "section": "xcmPallet",
  },
]
`;
